### PR TITLE
Audit Log PR Revisions

### PR DESCRIPTION
**Pull request by baonguyenNava**
## Summary

Revisions and enhancements to the audit log feature, including improvements to transactional event recording, documentation refinements, and expanded test coverage.

## Changes

- Enhancements to audit log event recording mechanism
- Refinements to documentation and PII redaction mechanism
- Audit log view improvements in dummy app
- Extended test coverage for edge cases and virtual actors
- Model, policy, and controller refinements

### CODE REVIEW

File: app/models/strata/application_form.rb [hunk 1]
```diff
@@ -23,7 +23,6 @@ class ApplicationForm < ApplicationRecord
 
     include Strata::Attributes
     include Strata::Determinable
-    include Strata::Auditable
```
Comment (baonguyenNava) [inline app/models/strata/application_form.rb:26]: Make Auditable opt-in

File: docs/decisions/audit-log-pii-redaction.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
+# Design Spec: PII redaction for `Strata::AuditLine#data`
```
Comment (baonguyenNava) [inline docs/decisions/audit-log-pii-redaction.md:1]: One concern that I have is about the possibility that PII makes it into the audit logs. This is a design spec that Claude created and that I reviewed as a possible implementation that we can take to avoid issues with PII being logged to the audit log.

File: docs/strata-audit-log.md [hunk 3]
```diff
@@ -105,6 +105,35 @@ The generator copies `db/migrate/<timestamp>_create_strata_audit_lines.rb`. It d
 
 Pass `--skip-migration-check` to suppress the post-generate prompt that asks whether to run `db:migrate` immediately.
 
+## Authorization
+
+Audit lines are polymorphic, which means the audit table mixes records belonging to different host models in one place. Hosts must wire authorization themselves — there is no global "you can see all audit lines" rule that's safe by default.
+
+The recommended pattern is a Pundit `Strata::AuditLinePolicy::Scope` that filters audit lines by joining each subject type to that subject's existing policy scope. The dummy app under `spec/dummy/app/policies/strata/audit_line_policy.rb` ships a reference implementation hosts can copy and adapt:
+
+```ruby
```
Comment (baonguyenNava) [inline docs/strata-audit-log.md:114]: Another issue that I raised is the fact that the audit logs are not scoped to the user. This would be a pattern that we can follow to adjust the scoping for audit logs.

File: docs/decisions/audit-log-pii-redaction.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+### Option B — Blacklist on the subject model
+
+Inverted form: hosts list the attributes they want *excluded*.
+
+```ruby
+class User < ApplicationRecord
+  include Strata::Auditable
+
+  audit_blacklist :password_digest, :ssn, :session_token
```
Comment (MichaelCrawfordNava) [inline docs/decisions/audit-log-pii-redaction.md]: Avoid usage of "blacklist" and prefer "blocklist" or another similar term

File: docs/decisions/audit-log-pii-redaction.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+- This is **not** a substitute for not putting secrets in `data`. Hosts that
+  pass arbitrary `request.params` will still leak — the goal here is to make
+  the structured-attribute path safe by default, not to police arbitrary
+  hashes.
+- This does not retroactively scrub existing audit lines. Hosts that need to
+  remediate historical PII must run their own data migration.
+
+## Design options considered
+
+### Option A — Whitelist on the subject model (proposed default)
```
Comment (MichaelCrawfordNava) [inline docs/decisions/audit-log-pii-redaction.md]: Avoid usage of "whitelist" and prefer "allowlist" or another similar term

File: spec/models/concerns/strata/auditable_spec.rb [hunk 6]
```diff
@@ -33,4 +33,10 @@
       expect { test_form.destroy }.not_to change(Strata::AuditLine, :count)
     end
   end
+
+  describe 'opt-in posture' do
+    it 'is not included in Strata::ApplicationForm by default' do
+      expect(Strata::ApplicationForm.include?(described_class)).to be(false)
+    end
+  end
```
Comment (MichaelCrawfordNava) [inline spec/models/concerns/strata/auditable_spec.rb]: Do we need this test? If so, we might want to include a test for Case too, just for coverage of "not included by default".

File: app/models/strata/audit_line.rb [hunk 7]
```diff
@@ -1,10 +1,16 @@
...
+  # AuditLine is an immutable record of something that happened.
+  #
+  # Columns:
+  # - `action` (string, required) — short event name, e.g. `"case.approved"`.
+  # - `subject_type` / `subject_id` (polymorphic, optional) — the record the
+  #   event is about.
+  # - `actor_type` / `actor_id` (polymorphic, optional) — who did it.
+  # - `data` (jsonb, defaults to `{}`) — free-form caller-supplied payload.
+  # - `created_at` (datetime) — when the line was recorded. There is no
+  #   `updated_at`; lines are immutable.
```
Comment (MichaelCrawfordNava) [inline app/models/strata/audit_line.rb]: It would probably be good to have these columns be attributes on the class. Or at least referenced with an `attr_reader` or something similar.

**Comment by baonguyenNava** [comment]
@MichaelCrawfordNava I address your feedback and updated the PR based on our discussion this morning.

File: app/models/strata/audit_line.rb [hunk 8]
```diff
@@ -19,11 +15,15 @@ class AuditLine < ApplicationRecord
     belongs_to :subject, polymorphic: true, optional: true
     belongs_to :actor,   polymorphic: true, optional: true
 
+    attribute :action,     :string
+    attribute :data,       :jsonb, default: {}
+    attribute :created_at, :datetime
```
Comment (MichaelCrawfordNava) [inline app/models/strata/audit_line.rb]: Might be worth making this readonly (since it's set automatically)

**Review by MichaelCrawfordNava — APPROVED** [review]

File: app/models/strata/virtual_actor.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+module Strata
```
Comment (baonguyenNava) [inline app/models/strata/virtual_actor.rb:3]: This is new: address when the actor is a system or API that changed the Active Record Object, instead of a user or another AR object acting on it.

File: app/models/strata/audit_line.rb [hunk 10]
```diff
@@ -19,15 +15,59 @@ class AuditLine < ApplicationRecord
...
+      end
+    }
+    scope :with_action,  ->(action)  { where(action: action.to_s) }
     scope :latest_first, -> { order(created_at: :desc) }
 
     def readonly?
       persisted?
     end
+
+    def actor
```
Comment (baonguyenNava) [inline app/models/strata/audit_line.rb:44]: Redefining the getter and setter for actor to account for a non-active record object.

File: app/models/strata/audit_line.rb (hunk 8 above)
Comment (baonguyenNava) [reply app/models/strata/audit_line.rb]: I removed it since it is unnecessary.

**Comment by baonguyenNava** [comment]
@MichaelCrawfordNava @jdettmannnava I added one more piece to this PR, which is to address an actor type that would account for a PORO, like API::Client in OSCER being the actor.

File: docs/decisions/audit-log-pii-redaction.md (hunk 2 above)
Comment (jdettmannnava) [reply docs/decisions/audit-log-pii-redaction.md:1]: Non-blocking comment: I am not understanding why we care if PII gets into the audit log? One of the uses of audit logs (not for oscer) is things like "changed address from a to b". Why is this a problem?
We can move this discussion to Slack if you think it is important enough to discuss more broadly.

I completely agree with the data not being a dump of anything that "might be useful some day" like request.params.

**Review by jdettmannnava — APPROVED** [review]
Great solution!
